### DISCUSSION TITLE
useAsTitle function hook

### DISCUSSION THREAD

**Discussion by katywings** [payloadcms/payload#1105] [Feature Requests & Ideas]
Maybe its a terrible idea for performance :thinking: , but I think it would be very useful if the `useAsTitle` config could accept a function taking the current document and returning any string as title. That way the developer could implement titles for many use cases.

Related:
- #808 
- #623

**Comment by jmikrut** [comment]
Yeah! I don't think this would be too bad for performance actually and we would happily accept a PR to allow for this. It'd be a pretty simple one to do. I'd probably just CMD+F for all instances of `useAsTitle` and then ensure each one is updated properly to suit the new API. Being that `useAsTitle` is only relevant in the admin UI, it should have no impact on the API side at all, making this potentially pretty straightforward.

Good call! I like it!

**Reply by katywings** [reply]
@jmikrut Thank you for your analysis! I thought a bit further about the implications of this. Depending on how we implement this I see two possible tradeoffs that we have to take:
1. Localization support: Do you think its needed for `useAsTitle`? Then we should write the `useAsTitle` function call like a react hook, so that `useLocale` could be called in `useAsTitle` :thinking: 
2. Async support: Are there use cases were the custom `useAsTitle` function would depend on async code? What is the general design concept on other similar functions in Payload?

**Comment by zubricks** [comment]
The team just reviewed this discussion and came up with a few reason why this will actually be quite a negative performance impact. It will also never be able to support async functions because of its heavy usage in the relationship field component. 

A really nice way to do this would be to use a disabled field and a field hook to generate a title on the backend. We actually do this in our own website CMS where we use a hidden `fullTitle` field that is generated on the backend with hooks.

**Reply by katywings** [reply]
@zubricks Thank you and the team for the feedback! I saw the approach with the field hook and I think it is very reasonable :+1:. 

Maybe one possible overall project-strategy in the long term, for similar questions could be a "best practices / faq" section in the docs :thinking:? E.g. "How can I show a combination of multiple fields as title?" -> "Set up a disabled field as useAsTitle and use a field hook to generate the title on save"

**Comment by VanCoding** [comment]
@zubricks I think to work around performance issues this could be split into 2 separate changes:
- Allow passing an array or something to "useAsTitle" that allows fetching multiple fields to use as title
- Allow overriding the "title" component that receives all those values specified in useAsTitle and renders them to a title value

What do you think?

The second change would be useful on its own, for cases where the title value is an object and it should be rendered to a title.
